### PR TITLE
dispatcher: split into multiple interfaces to improve functional scoping

### PR DESCRIPTION
**Pull request by goaway**
Commit Message: dispatcher: split into multiple interfaces to improve functional scoping
Additional Description: The Envoy event dispatcher is a bit of a multi-tool, and often only a small portion of its interface is used in a given context. Envoy Mobile has various cases where it needs to wrap or extend pieces of the Event::Dispatcher, but it's difficult to simply wrap it in its entirety given the amount of functionality it has. Having it implement more narrowly scoped interfaces makes it easier to do this.

This can also simplify mocking and testing dispatcher functionality in different contexts and could eventually support more compartmentalized implementation as well.
Risk Level: Moderate
Testing: Local & CI

### CODE REVIEW

**Comment by goaway** [comment]
Note this is straight refactor; it only moves the function declarations between interfaces, all of which the Dispatcher still implements. I did a search for the previous DispatcherBase and didn't find any usage.

File: source/common/common/scope_tracker.h [hunk 1]
```diff
@@ -10,20 +10,20 @@ namespace Envoy {
...
 // work done in this thread.
 //
-// When created, it appends the tracked object to the dispatcher's stack of tracked objects, and
-// when destroyed it pops the dispatcher's stack of tracked object, which should be the object it
+// When created, it appends the tracked object to the tracker's stack of tracked objects, and
+// when destroyed it pops the tracker's stack of tracked object, which should be the object it
 // registered.
 class ScopeTrackerScopeState {
 public:
   ScopeTrackerScopeState(const ScopeTrackedObject* object, Event::Dispatcher& dispatcher)
```
Comment (junr03) [inline source/common/common/scope_tracker.h:18]:
```suggestion
  ScopeTrackerScopeState(const ScopeTrackedObject* object, Event::ScopeTracker& tracker)
```

**Comment by junr03** [comment]
@antoniovicente did you have additional feedback here?

@goaway do you mind merging main?

**Review by antoniovicente — APPROVED** [review]
Looks good.

Sorry for the delays in review.  I have been moving and that has required being OOO for a few days.

**Comment by goaway** [comment]
Thanks @antoniovicente!

**Pull request merged**